///////////////////////////////////////////////////
//         Define Protobuf Syntax Version        //         
///////////////////////////////////////////////////
syntax = "proto3";

///////////////////////////////////////////////////
//              Define Package Name              //         
///////////////////////////////////////////////////
package spi;

///////////////////////////////////////////////////
//               Define RPC Service              //         
///////////////////////////////////////////////////
service SPI {
    // Configure DAC Channel 0
    rpc ConfigureDACChannel0(ConfigureDACChannel0Packet) returns (ConfigureDACChannel0Response) {}
    // Configure DAC Channel 1
    rpc ConfigureDACChannel1(ConfigureDACChannel1Packet) returns (ConfigureDACChannel1Response) {}
    // Configure DAC Channel 2
    rpc ConfigureDACChannel2(ConfigureDACChannel2Packet) returns (ConfigureDACChannel2Response) {}
    // Read ADC
    rpc ReadADC(ReadADCPacket) returns (ReadADCResponse) {}
}

///////////////////////////////////////////////////
//              Define Message Type              //         
///////////////////////////////////////////////////
// Configure DAC Channel 0 Message
message ConfigureDACChannel0Packet {
    // Define the fields
    string channel_code = 1;
}

// Configure DAC Channel 0 Response
message ConfigureDACChannel0Response {
    // Define the fields
    string operation_name = 1;
    string start_time = 2;
    string end_time = 3;
    bool success = 4;
}

// Configure DAC Channel 1 Message
message ConfigureDACChannel1Packet {
    // Define the fields
    string channel_code = 1;
}

// Configure DAC Channel 1 Response
message ConfigureDACChannel1Response {
    // Define the fields
    string operation_name = 1;
    string start_time = 2;
    string end_time = 3;
    bool success = 4;
}

// Configure DAC Channel 2 Message
message ConfigureDACChannel2Packet {
    // Define the fields
    string channel_code = 1;
}

// Configure DAC Channel 2 Response
message ConfigureDACChannel2Response {
    // Define the fields
    string operation_name = 1;
    string start_time = 2;
    string end_time = 3;
    bool success = 4;
}

// Configure Read ADC Message
message ReadADCPacket {}

// Configure Read ADC Response
message ReadADCResponse {
    // Define the fields
    string operation_name = 1;
    repeated int32 adc_value = 2;
    repeated int32 adc_id = 3;
    repeated int32 adc_softspan = 4;
    repeated float adc_voltage = 5;
    string start_time = 6;
    string end_time = 7;
    bool success = 8;
}
